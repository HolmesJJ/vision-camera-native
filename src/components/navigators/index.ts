/* eslint-disable prettier/prettier */
// GENERATED CODE: DO NOT EDIT!

import { RouteDefinition } from './RouteDefinition';
import { StackParamList } from './StackParamList';

export {
  RouteDefinition,
  StackParamList,
};
