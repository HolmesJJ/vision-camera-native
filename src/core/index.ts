/* eslint-disable prettier/prettier */
// GENERATED CODE: DO NOT EDIT!

import {
  ApplicationContext,
  useApplicationContext,
} from './ApplicationContext';
import * as BuildConfig from './BuildConfig';
import * as mocks from './mocks';
import * as utilities from './utilities';

export {
  ApplicationContext,
  BuildConfig,
  mocks,
  useApplicationContext,
  utilities,
};
