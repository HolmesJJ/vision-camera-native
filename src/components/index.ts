/* eslint-disable prettier/prettier */
// GENERATED CODE: DO NOT EDIT!

import * as navigators from './navigators';

export {
  navigators,
};
