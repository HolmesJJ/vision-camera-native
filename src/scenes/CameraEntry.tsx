/**
 * CameraEntry.tsx
 * @author HU JIAJUN <e0556371@u.nus.edu>
 * @file Entry point for Camera Scene
 */

import React from 'react';
import { AppState, AppStateStatus, StatusBar, View } from 'react-native';
import { SafeAreaView } from 'react-native-safe-area-context';
import {
  Camera,
  CameraDevice,
  CameraDevices,
  useCameraDevices,
  CameraRuntimeError,
  useFrameProcessor,
  FrameProcessorPerformanceSuggestion,
  Frame,
} from 'react-native-vision-camera';
import { useIsFocused } from '@react-navigation/core';
import * as Colors from 'styles/Colors';
import { Option } from 'nasi-lemak';

import { RouteDefinition, StackScreenProps } from 'components/navigators';
import { useCameraEntry as useStyles } from 'styles/camera';

export interface ICameraEntryProps
  extends StackScreenProps<RouteDefinition.CAMERA> {}

export function CameraEntry(_props: ICameraEntryProps) {
  const styles = useStyles();

  const onError = React.useCallback((error: CameraRuntimeError) => {
    console.error(error);
  }, []);

  // Check if camera page is active
  const [isForeground, setIsForeground] = React.useState<boolean>(true);
  React.useEffect(() => {
    const onChange = (state: AppStateStatus): void => {
      setIsForeground(state === 'active');
    };
    const listener = AppState.addEventListener('change', onChange);
    return () => listener.remove();
  }, [setIsForeground]);
  const isFocussed: boolean = useIsFocused();
  const isActive: boolean = isFocussed && isForeground;

  // Camera Format Settings
  const devices: CameraDevices = useCameraDevices();
  const device: CameraDevice | undefined = devices.back;

  const frameProcessor = useFrameProcessor((frame: Frame) => {
    'worklet';
    console.log('Width: ' + frame.width + ', Height: ' + frame.height);
  }, []);

  const onFrameProcessorSuggestionAvailable = React.useCallback(
    (suggestion: FrameProcessorPerformanceSuggestion) => {
      console.log(
        `Suggestion available! ${suggestion.type}: Can do ${suggestion.suggestedFrameProcessorFps} FPS`,
      );
    },
    [],
  );

  return (
    <SafeAreaView style={styles.safeArea}>
      <StatusBar backgroundColor={Colors.black} barStyle="light-content" />
      <View style={styles.container}>
        {Option.isSome(device) && (
          <Camera
            style={styles.camera}
            device={device!}
            isActive={isActive}
            onError={onError}
            frameProcessor={frameProcessor}
            orientation="portrait"
            onFrameProcessorPerformanceSuggestionAvailable={
              onFrameProcessorSuggestionAvailable
            }
          />
        )}
      </View>
    </SafeAreaView>
  );
}
