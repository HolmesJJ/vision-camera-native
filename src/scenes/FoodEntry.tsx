/**
 * FoodEntry.tsx
 * @author HU JIAJUN <e0556371@u.nus.edu>
 * @file Entry point for Food Scene
 */

import React from 'react';
import { Image, StatusBar, View } from 'react-native';
import { SafeAreaView } from 'react-native-safe-area-context';
import { Button, Text, TextInput } from 'react-native-paper';
import 'react-native-reanimated';
import * as Colors from 'styles/Colors';

import { RouteDefinition, StackScreenProps } from 'components/navigators';
import { useFoodEntry as useStyles } from 'styles/food';

export interface IFoodEntryProps
  extends StackScreenProps<RouteDefinition.FOOD> {}

export function FoodEntry(_props: IFoodEntryProps) {
  const styles = useStyles();
  const [id, setId] = React.useState('075779311145');
  const [productInfo, setProductInfo] = React.useState({
    imageUrl: '',
    energyKcal: '',
    energyKj: '',
    fat: '',
    carbohydrates: '',
    sugars: '',
    fiber: '',
    proteins: '',
    salt: '',
    ingredients: '',
  });

  const fetchProductData = async () => {
    try {
      const response = await fetch(
        `https://world.openfoodfacts.net/api/v2/product/${id}`,
      );
      const data = await response.json();
      const { product } = data;
      setProductInfo({
        imageUrl: product.image_url,
        energyKcal: product.nutriments['energy-kcal']
          ? product.nutriments['energy-kcal'].toFixed(2)
          : 'N/A',
        energyKj: product.nutriments.energy
          ? product.nutriments.energy.toFixed(2)
          : 'N/A',
        fat: product.nutriments.fat ? product.nutriments.fat.toFixed(2) : 'N/A',
        carbohydrates: product.nutriments.carbohydrates
          ? product.nutriments.carbohydrates.toFixed(2)
          : 'N/A',
        sugars: product.nutriments.sugars
          ? product.nutriments.sugars.toFixed(2)
          : 'N/A',
        fiber: product.nutriments.fiber
          ? product.nutriments.fiber.toFixed(2)
          : 'N/A',
        proteins: product.nutriments.proteins
          ? product.nutriments.proteins.toFixed(2)
          : 'N/A',
        salt: product.nutriments.salt
          ? product.nutriments.salt.toFixed(2)
          : 'N/A',
        ingredients: product.ingredients_text
          ? product.ingredients_text
          : 'N/A',
      });
    } catch (error) {
      console.error('Error fetching product data:', error);
    }
  };

  return (
    <SafeAreaView style={styles.safeArea}>
      <StatusBar backgroundColor={Colors.black} barStyle="light-content" />
      <View style={styles.container}>
        <TextInput
          label="Product ID"
          value={id}
          onChangeText={(text) => setId(text)}
          mode="outlined"
          style={styles.input}
        />
        <Button
          mode="contained"
          onPress={fetchProductData}
          style={styles.button}
        >
          CHECK
        </Button>
        {productInfo.imageUrl ? (
          <Image source={{ uri: productInfo.imageUrl }} style={styles.image} />
        ) : null}
        <View style={styles.tableContainer}>
          <View style={styles.tableRow}>
            <Text style={styles.tableLabel}>Energy (kcal):</Text>
            <Text style={styles.tableValue}>
              {productInfo.energyKcal} kcal/100g
            </Text>
          </View>
          <View style={styles.tableRow}>
            <Text style={styles.tableLabel}>Energy (kJ):</Text>
            <Text style={styles.tableValue}>
              {productInfo.energyKj} kJ/100g
            </Text>
          </View>
          <View style={styles.tableRow}>
            <Text style={styles.tableLabel}>Fat:</Text>
            <Text style={styles.tableValue}>{productInfo.fat} g/100g</Text>
          </View>
          <View style={styles.tableRow}>
            <Text style={styles.tableLabel}>Carbohydrates:</Text>
            <Text style={styles.tableValue}>
              {productInfo.carbohydrates} g/100g
            </Text>
          </View>
          <View style={styles.tableRow}>
            <Text style={styles.tableLabel}>Sugars:</Text>
            <Text style={styles.tableValue}>{productInfo.sugars} g/100g</Text>
          </View>
          <View style={styles.tableRow}>
            <Text style={styles.tableLabel}>Fiber:</Text>
            <Text style={styles.tableValue}>{productInfo.fiber} g/100g</Text>
          </View>
          <View style={styles.tableRow}>
            <Text style={styles.tableLabel}>Proteins:</Text>
            <Text style={styles.tableValue}>{productInfo.proteins} g/100g</Text>
          </View>
          <View style={styles.tableRow}>
            <Text style={styles.tableLabel}>Salt:</Text>
            <Text style={styles.tableValue}>{productInfo.salt} g/100g</Text>
          </View>
          <View style={styles.tableRow}>
            <Text style={styles.tableLabel}>Ingredients:</Text>
            <Text style={styles.tableValue}>{productInfo.ingredients}</Text>
          </View>
        </View>
      </View>
    </SafeAreaView>
  );
}
