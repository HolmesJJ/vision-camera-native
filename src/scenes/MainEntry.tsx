/**
 * MainEntry.tsx
 * @author HU JIAJUN <e0556371@u.nus.edu>
 * @file Entry point for Main Scene
 */

import React from 'react';
import { NativeModules, StatusBar, View, Linking } from 'react-native';
import { Button, Text } from 'react-native-paper';
import { SafeAreaView } from 'react-native-safe-area-context';
import { Camera, CameraPermissionStatus } from 'react-native-vision-camera';
import { Optional } from 'nasi-lemak';

import { RouteDefinition, StackScreenProps } from 'components/navigators';
import { useMainEntry as useStyles } from 'styles/main';

export interface IMainEntryProps
  extends StackScreenProps<RouteDefinition.MAIN> {}

export function MainEntry(props: IMainEntryProps) {
  const navigation = props.navigation;
  const { ActivityStarter } = NativeModules;
  const styles = useStyles();

  const [cameraPermissionStatus, setCameraPermissionStatus] =
    React.useState<CameraPermissionStatus>('not-determined');
  const [microphonePermissionStatus, setMicrophonePermissionStatus] =
    React.useState<CameraPermissionStatus>('not-determined');
  const [disabled, setDisabled] = React.useState<boolean>(true);

  const requestMicrophonePermission = React.useCallback(async () => {
    const permission = await Camera.requestMicrophonePermission();
    if (permission === 'denied') {
      await Linking.openSettings();
    }
    setMicrophonePermissionStatus(permission);
  }, []);

  const requestCameraPermission = React.useCallback(async () => {
    const permission = await Camera.requestCameraPermission();
    if (permission === 'denied') {
      await Linking.openSettings();
    }
    setCameraPermissionStatus(permission);
  }, []);

  React.useEffect(() => {
    setCameraPermissionStatus(Camera.getCameraPermissionStatus());
    setMicrophonePermissionStatus(Camera.getMicrophonePermissionStatus());
  }, []);

  React.useEffect(() => {
    if (
      cameraPermissionStatus === 'granted' &&
      microphonePermissionStatus === 'granted'
    ) {
      setDisabled(false);
    }
  }, [cameraPermissionStatus, microphonePermissionStatus, navigation]);

  return (
    <SafeAreaView style={styles.safeArea}>
      <StatusBar
        translucent
        backgroundColor="transparent"
        barStyle="dark-content"
      />
      <View style={styles.container}>
        <Optional predicate={disabled}>
          <View>
            {cameraPermissionStatus !== 'granted' && (
              <View style={styles.permissionContainer}>
                <Text style={styles.permissionText}>
                  Vision Camera needs camera permission
                </Text>
                <Button
                  style={styles.permissionButton}
                  testID="cameraButton"
                  mode="contained"
                  onPress={requestCameraPermission}
                >
                  Grant
                </Button>
              </View>
            )}
            {microphonePermissionStatus !== 'granted' && (
              <View style={styles.permissionContainer}>
                <Text style={styles.permissionText}>
                  Vision Camera needs microphone permission
                </Text>
                <Button
                  style={styles.permissionButton}
                  testID="cameraButton"
                  mode="contained"
                  onPress={requestMicrophonePermission}
                >
                  Grant
                </Button>
              </View>
            )}
          </View>
        </Optional>
        <Optional predicate={!disabled}>
          <Button
            style={styles.cameraButton}
            testID="visionCameraButton"
            mode="contained"
            disabled={disabled}
            onPress={() => {
              navigation.navigate(RouteDefinition.CAMERA);
            }}
          >
            VISION CAMERA
          </Button>
        </Optional>
        <Optional predicate={!disabled}>
          <Button
            style={styles.cameraButton}
            testID="nativeCameraButton"
            mode="contained"
            disabled={disabled}
            onPress={() => ActivityStarter.navigateToCamera()}
          >
            NATIVE CAMERA
          </Button>
        </Optional>
        <Optional predicate={!disabled}>
          <Button
            style={styles.testButton}
            testID="testButton"
            mode="contained"
            disabled={disabled}
            onPress={() => {
              navigation.navigate(RouteDefinition.FOOD);
            }}
          >
            OPEN FOOD FACTS
          </Button>
        </Optional>
        <Optional predicate={!disabled}>
          <Button
            style={styles.testButton}
            testID="testButton"
            mode="contained"
            disabled={disabled}
            onPress={() => {
              navigation.navigate(RouteDefinition.TEST);
            }}
          >
            TEST MODEL
          </Button>
        </Optional>
      </View>
    </SafeAreaView>
  );
}
