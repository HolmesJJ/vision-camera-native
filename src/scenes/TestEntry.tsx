/**
 * TestEntry.tsx
 * @author HU JIAJUN <e0556371@u.nus.edu>
 * @file Entry point for Test Scene
 */

import React from 'react';
import { Image, StatusBar, View } from 'react-native';
import { Button, Text } from 'react-native-paper';
import { SafeAreaView } from 'react-native-safe-area-context';
import {
  Tensor,
  TensorflowModel,
  useTensorflowModel,
} from 'react-native-fast-tflite';
import { readFile } from 'react-native-fs';
import 'react-native-reanimated';
import * as Colors from 'styles/Colors';

import { RouteDefinition, StackScreenProps } from 'components/navigators';
import { useTestEntry as useStyles } from 'styles/test';

export interface ITestEntryProps
  extends StackScreenProps<RouteDefinition.TEST> {}

function tensorToString(tensor: Tensor): string {
  return `\n  - ${tensor.dataType} ${tensor.name}[${tensor.shape}]`;
}

function modelToString(model: TensorflowModel): string {
  return (
    `TFLite Model (${model.delegate}):\n` +
    `- Inputs: ${model.inputs.map(tensorToString).join('')}\n` +
    `- Outputs: ${model.outputs.map(tensorToString).join('')}`
  );
}

export function TestEntry(_props: ITestEntryProps) {
  const styles = useStyles();

  // https://github.com/mrousavy/react-native-fast-tflite/issues/11
  React.useEffect(() => {
    const processImage = async () => {
      console.log('aaa');
      const imagePath = 'assets/images/cat.jpg';
      const imageBase64 = await readFile(imagePath, 'base64').catch((err) =>
        console.error('Failed to read file:', err),
      );
      console.log(imageBase64);
    };
    processImage();
  }, []);

  const objectDetection = useTensorflowModel(
    require('assets/efficientdet.tflite'),
  );
  const model =
    objectDetection.state === 'loaded' ? objectDetection.model : undefined;

  React.useEffect(() => {
    if (model === undefined) {
      return;
    }
    console.log(`Model loaded! Shape:\n${modelToString(model)}]`);
  }, [model]);

  return (
    <SafeAreaView style={styles.safeArea}>
      <StatusBar backgroundColor={Colors.black} barStyle="light-content" />
      <View style={styles.container}>
        <Image source={require('assets/images/cat.jpg')} style={styles.image} />
        <Button
          style={styles.detectButton}
          testID="detectButton"
          mode="contained"
          onPress={() => {}}
        >
          Detect
        </Button>
        <Text style={styles.outputsText}>Results</Text>
      </View>
    </SafeAreaView>
  );
}
