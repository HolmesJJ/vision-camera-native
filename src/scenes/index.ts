/* eslint-disable prettier/prettier */
// GENERATED CODE: DO NOT EDIT!

import { CameraEntry } from './CameraEntry';
import { FoodEntry } from './FoodEntry';
import { MainEntry } from './MainEntry';
import { TestEntry } from './TestEntry';

export {
  CameraEntry,
  FoodEntry,
  MainEntry,
  TestEntry,
};
