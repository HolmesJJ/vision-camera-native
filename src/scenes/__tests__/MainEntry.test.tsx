/**
 * MainEntry.test.tsx
 * @author HU JIAJUN <e0556371@u.nus.edu>
 * @file Tests for MainEntry
 */

import React from 'react';
import { render } from '@testing-library/react-native';
import * as ReactNavigationNative from '@react-navigation/native';

import { RouteDefinition, StackParamList } from 'components/navigators';
import { MockStackScreenProps } from 'core/mocks';
import { MainEntry } from 'scenes/';

jest.useFakeTimers();

jest.mock('react-native-config', () => ({
  ENDPOINT_SERVER_URL: 'mock endpoint',
  ENDPOINT_POLL_INTERVAL_MS: 1000,
}));

const mockedNavigate = jest.fn();

jest.mock('@react-navigation/native', () => {
  return {
    ...(jest.requireActual(
      '@react-navigation/native',
    ) as typeof ReactNavigationNative),
    useNavigation: () => ({
      navigate: mockedNavigate,
    }),
  };
});

const props = new MockStackScreenProps<StackParamList, RouteDefinition.MAIN>();

test('snapshot', () => {
  const tree = render(<MainEntry {...props} />);
  expect(tree.toJSON()).toMatchSnapshot();
});
