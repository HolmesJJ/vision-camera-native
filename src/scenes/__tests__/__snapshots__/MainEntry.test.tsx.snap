// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<RNCSafeAreaView
  edges={
    Object {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "center",
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "marginBottom": 50,
            "marginLeft": 20,
            "marginRight": 20,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "textAlign": "left",
              },
              Object {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontWeight": "400",
                "letterSpacing": 0,
              },
              Object {
                "writingDirection": "ltr",
              },
              Object {
                "fontSize": 18,
                "marginBottom": 10,
                "textAlign": "center",
              },
            ]
          }
        >
          Vision Camera needs camera permission
        </Text>
        <View
          collapsable={false}
          style={
            Object {
              "backgroundColor": "rgba(103, 80, 164, 1)",
              "borderRadius": 20,
              "shadowColor": "#000",
              "shadowOffset": Object {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
              "width": 100,
            }
          }
          testID="cameraButton-container-outer-layer"
        >
          <View
            collapsable={false}
            style={
              Object {
                "backgroundColor": "rgba(103, 80, 164, 1)",
                "borderColor": "transparent",
                "borderRadius": 20,
                "borderStyle": "solid",
                "borderWidth": 0,
                "flex": undefined,
                "minWidth": 64,
                "shadowColor": "#000",
                "shadowOffset": Object {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0,
                "shadowRadius": 0,
              }
            }
            testID="cameraButton-container"
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                Object {
                  "disabled": false,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  Object {
                    "overflow": "hidden",
                  },
                  Object {
                    "borderRadius": 20,
                  },
                ]
              }
              testID="cameraButton"
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "center",
                    },
                    undefined,
                  ]
                }
              >
                <Text
                  numberOfLines={1}
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "textAlign": "left",
                      },
                      Object {
                        "color": "rgba(28, 27, 31, 1)",
                        "writingDirection": "ltr",
                      },
                      Array [
                        Object {
                          "fontFamily": "System",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "letterSpacing": 0.1,
                          "lineHeight": 20,
                        },
                        Array [
                          Object {
                            "marginHorizontal": 16,
                            "marginVertical": 9,
                            "textAlign": "center",
                          },
                          false,
                          Object {
                            "marginHorizontal": 24,
                            "marginVertical": 10,
                          },
                          undefined,
                          false,
                          Object {
                            "color": "rgba(255, 255, 255, 1)",
                            "fontFamily": "System",
                            "fontSize": 14,
                            "fontWeight": "500",
                            "letterSpacing": 0.1,
                            "lineHeight": 20,
                          },
                          undefined,
                        ],
                      ],
                    ]
                  }
                  testID="cameraButton-text"
                >
                  Grant
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "marginBottom": 50,
            "marginLeft": 20,
            "marginRight": 20,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "textAlign": "left",
              },
              Object {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontWeight": "400",
                "letterSpacing": 0,
              },
              Object {
                "writingDirection": "ltr",
              },
              Object {
                "fontSize": 18,
                "marginBottom": 10,
                "textAlign": "center",
              },
            ]
          }
        >
          Vision Camera needs microphone permission
        </Text>
        <View
          collapsable={false}
          style={
            Object {
              "backgroundColor": "rgba(103, 80, 164, 1)",
              "borderRadius": 20,
              "shadowColor": "#000",
              "shadowOffset": Object {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
              "width": 100,
            }
          }
          testID="cameraButton-container-outer-layer"
        >
          <View
            collapsable={false}
            style={
              Object {
                "backgroundColor": "rgba(103, 80, 164, 1)",
                "borderColor": "transparent",
                "borderRadius": 20,
                "borderStyle": "solid",
                "borderWidth": 0,
                "flex": undefined,
                "minWidth": 64,
                "shadowColor": "#000",
                "shadowOffset": Object {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0,
                "shadowRadius": 0,
              }
            }
            testID="cameraButton-container"
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                Object {
                  "disabled": false,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  Object {
                    "overflow": "hidden",
                  },
                  Object {
                    "borderRadius": 20,
                  },
                ]
              }
              testID="cameraButton"
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "center",
                    },
                    undefined,
                  ]
                }
              >
                <Text
                  numberOfLines={1}
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "textAlign": "left",
                      },
                      Object {
                        "color": "rgba(28, 27, 31, 1)",
                        "writingDirection": "ltr",
                      },
                      Array [
                        Object {
                          "fontFamily": "System",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "letterSpacing": 0.1,
                          "lineHeight": 20,
                        },
                        Array [
                          Object {
                            "marginHorizontal": 16,
                            "marginVertical": 9,
                            "textAlign": "center",
                          },
                          false,
                          Object {
                            "marginHorizontal": 24,
                            "marginVertical": 10,
                          },
                          undefined,
                          false,
                          Object {
                            "color": "rgba(255, 255, 255, 1)",
                            "fontFamily": "System",
                            "fontSize": 14,
                            "fontWeight": "500",
                            "letterSpacing": 0.1,
                            "lineHeight": 20,
                          },
                          undefined,
                        ],
                      ],
                    ]
                  }
                  testID="cameraButton-text"
                >
                  Grant
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaView>
`;
